### PR TITLE
machine/flash: improvements and fixes to FlashBuffer Write() and Seek()

### PR DESCRIPTION
**Pull request by deadprogram**
This PR has improvements and fixes to the `FlashBuffer` `Write()` and `Seek()` functions.

It also updates the example to show that you can perform multiple `Write()` operations, as long as the data is a multiple of the write block size.

### CODE REVIEW

**Comment by deadprogram** [comment]
@sago35 I think this PR should address your issues, with the buffer size limitations mentioned above.

**Comment by deadprogram** [comment]
```
$ tinygo monitor
Connected to /dev/ttyACM0. Press Ctrl-C to exit.
Flash data start:       0x10003000
Flash data end:         0x10200000
Flash data size, bytes: 2084864
Flash write block size: 256
Flash erase block size: 4096

Reading data from flash: 
1234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321
1234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321
1234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321
1234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321
1234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321
1234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321
1234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321
1234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321
1234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321
1234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321
Writing data to flash: 
1234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321
1234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321
1234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321
1234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321
1234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321
1234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321
1234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321
1234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321
1234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321
1234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321
Reading data back from flash: 
1234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321
1234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321
1234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321
1234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321
1234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321
1234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321
1234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321
1234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321
1234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321
1234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321123456788765432112345678876543211234567887654321
Done.
```

File: src/machine/flash.go [hunk 1]
```diff
@@ -133,12 +149,16 @@ func (fl *FlashBuffer) Close() error {
...
 // Seek implements io.Seeker interface, but with limitations.
 // You can only seek relative to the start.
-// Also, you cannot use seek before write operations, only read.
+// Also, with write operations you can only seek back to the beginning.
 func (fl *FlashBuffer) Seek(offset int64, whence int) (int64, error) {
 	if whence != io.SeekStart {
 		panic("you can only Seek relative to Start")
 	}
 
+	if offset < 0 || uintptr(offset) >= FlashDataEnd() {
```
Comment (sago35) [inline src/machine/flash.go:158]: Perhaps ` offset < End - Start` is the correct value.

File: src/machine/flash.go [hunk 2]
```diff
@@ -89,38 +93,50 @@ func (fl *FlashBuffer) Read(p []byte) (n int, err error) {
...
 	// block 0 -> fl.start
-	// block 1 -> fl.start + pagesize
-	// block 2 -> fl.start + pagesize*2
+	// block 1 -> fl.start + blocksize
+	// block 2 -> fl.start + blocksize*2
 	// ...
-	currentPageBlock := (fl.start + fl.offset - FlashDataStart()) + (pagesize-1)/pagesize
-	lastPageBlockNeeded := (fl.start + fl.offset + uintptr(len(p)) - FlashDataStart()) + (pagesize-1)/pagesize
+	currentPageBlock := (fl.offset + (blocksize - 1)) / blocksize
+	lastPageBlockNeeded := (fl.offset + uintptr(len(p)) + (blocksize - 1)) / blocksize
```
Comment (sago35) [inline src/machine/flash.go:111]: Why add (blocksize-1) here?
If fl.offset is 0, the correct value 0 is returned; if fl.offset is 1, 1 is returned. But isn't it correct to return 0 even when fl.offset is 1?

File: src/machine/flash.go [hunk 3]
```diff
@@ -89,38 +93,50 @@ func (fl *FlashBuffer) Read(p []byte) (n int, err error) {
...
+	}
 
 	// erase enough blocks to hold the data
-	if err := fl.b.EraseBlocks(int64(currentPageBlock), int64(lastPageBlockNeeded-currentPageBlock)); err != nil {
-		return 0, err
+	if lastPageBlockNeeded*blocksize > fl.erased {
+		if err := fl.b.EraseBlocks(int64(currentPageBlock), int64(lastPageBlockNeeded-currentPageBlock)); err != nil {
+			return 0, err
+		}
+		fl.erased = lastPageBlockNeeded * blocksize
```
Comment (sago35) [inline src/machine/flash.go:125]: I believe this approach will not work.
For example, when using the filesystem, the areas are not used in order from the beginning to the end.
It is also very confusing in combination with Seek.

I guess the only way is to read, erase, rewrite, and then write.

Comment (deadprogram) [reply src/machine/flash.go:125]: > I guess the only way is to read, erase, rewrite, and then write.

I think that is probably true when using the `FlashBuffer`. It is just a shortcut? Perhaps it is not really complete enough as @aykevl suggests we should remove it.

For a real file system, try out https://github.com/tinygo-org/tinyfs/tree/dev which uses the `BlockDevice` interface.

Comment (sago35) [reply src/machine/flash.go:125]: If so, how about the following?
`EraseBlocks()` are a little more difficult to handle.
And even if not necessarily using tinyfs, there may be cases where the user wants to save a little data.

* Write does not delete, but simply writes
* Read simply reads
* Create a function that returns the block number contained in the specified address to assist EraseBlock
    * ex) func XXXXX(addr uintptr) (blockNo, startAddr)

**Comment by deadprogram** [comment]
See https://github.com/tinygo-org/tinygo/pull/3718 for a pure `BlockDevice` implementation.

**Comment by deadprogram** [comment]
Closing since replaced by #3718

**Pull request closed without merging**